Implemented json models hooks

Elix-x opened the pull request:
## What did you do _O_o_ ???

Created a system that allows extension of vanilla model json format by modders in order to add functionality or change existing one. Allows to affect baked result on any level (from textures to vertex data). Capabilities are used to as data holders on different levels, they are also deserialized from json. Events are used in order to affect baked outcome, with access to capabilities to take decisions based on original json.

```
JSON ==Deserialization====> Model --------{     Parts for each depth level     } ===> Baking
                       ||                 {                 |                           ||=====> Baked Model
                       ===> Capabilities -{  Capabilities for each depth level } ===> Events
```
## Examples

I included 2 examples with this PR. One for model level modification and another one for vertex level modification.
[Here's example mod with capabilities and events](https://github.com/elix-x/MinecraftForge/blob/8eb94cf1af76a713610a3cc9e86ace6a8fa8ec24/src/test/java/net/minecraftforge/test/JsonModelCapabilityTest.java), [here the model file](https://github.com/elix-x/MinecraftForge/blob/8eb94cf1af76a713610a3cc9e86ace6a8fa8ec24/src/test/resources/assets/jsonmodelcapabilitytest/models/block/testblock.json) and here the result:
![Fancy](http://i.imgur.com/Llojd7z.png)

_**Note: these are the actual functionalities i need for a mod.**_
### Include:

Includes all quads of another model into this one with possible transformations. Does not require a ton of copy-pasta for repeating complex models.
Result:
![Bacon!](http://i.imgur.com/j80IHDf.png)
Code (responsible parts of code of example mod, cut out for readability):

``` java
    @CapabilityInject(Include.class)
    private static final Capability<Include> include = null;
...
    CapabilityManager.INSTANCE.register(Include.class, new Include.Storage(), Include.class);
...
    @SubscribeEvent
    public void attachCapsModel(ClientAttachCapabilitiesEvent.ModelBlock event){
        event.addCapability(new ResourceLocation("test:include"), new ICapabilitySerializable<NBTBase>(){...});
    }
...
    @SubscribeEvent
    public void include(VanillaModelWrapperEvent.Bake.Model.Post event){
        if(event.getModel().hasCapability(include, null)){
            ...
        }
    }
...
    public static class Include {

        List<Pair<String, ComplexTransforms>> include = new ArrayList<Pair<String,ComplexTransforms>>();

        static class Storage implements IStorage<Include> { ... }

        }
    public static class ComplexTransforms
    {

        List<Transform> transforms = new ArrayList<Transform>();

        ...

        static class Transform {

            private Vector3f translation = null;
            private Vector3f rotation = null;
            private Vector3f scale = null;

            ...

        }
    }
```

Example json:

``` json
{
    "capabilities":{
        "test:include": [
            {
                "model": "block/furnace",
                "transform": [
                    {
                        "rotation": {
                            "x": 0,
                            "y": 0,
                            "z": 180
                        },
                        "scale": {
                            "x": 0.5,
                            "y": 0.5,
                            "z": 0.5
                        }
                    }
                ]
            },
            {
                "model": "block/beacon",
                "transform": [
                    {
                        "scale": {
                            "x": 0.5,
                            "y": 0.5,
                            "z": 0.5
                        }
                    },
                    {
                        "translation": {
                            "x": 0,
                            "y": 1,
                            "z": 0
                        }
                    }
                ]
            },
            {
                "model": "block/end_rod",
                "transform": [
                    {
                        "translation": {
                        "x": 0,
                            "y": 1.25,
                            "z": 0
                        }
                    }
                ]
            }
        ]
    }
}
```
### UVLightmap

We have `shade`, but we can't cofigure intensity.
Result:
![LEDs!](http://i.imgur.com/bKEHaF9.png)
Code (responsible parts of code of example mod, cut out for readability):

``` java
    @CapabilityInject(UVLightmap.class)
    private static final Capability<UVLightmap> uvlightmap = null;
...
    CapabilityManager.INSTANCE.register(UVLightmap.class, new UVLightmap.Storage(), UVLightmap.class);
...
    @SubscribeEvent
    public void attachCapsBlockPartFace(ClientAttachCapabilitiesEvent.BlockPartFace event){
        event.addCapability(new ResourceLocation("test:uvlightmap"), new ICapabilitySerializable<NBTBase>() { ... });
    }
...
    @SubscribeEvent
    public void uvlightmap(BakeBlockPartFaceEvent event){
        if(event.getFace().hasCapability(uvlightmap, null)){
            ...
        }
    }
...
    public static class UVLightmap {

        float block = -1;
        float sky = -1;

        static class Storage implements IStorage<UVLightmap> {

            ...

        }

    }
```

Example json:

``` json
{
    "textures":{
        "all": "jsonmodelcapabilitytest:blocks/lights"
    },
    "elements": [
        {   
            "from": [ 2, 2, 2 ],
            "to": [ 14, 14, 14 ],
            "faces": {
                "down":  { "texture": "all", "cullface": "down", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.007, "sky": 0.007 } } },
                "up":    { "texture": "all", "cullface": "up", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.007, "sky": 0.007 } } },
                "north": { "texture": "all", "cullface": "north", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.004, "sky": 0.004 } } },
                "south": { "texture": "all", "cullface": "south", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } },
                "west":  { "texture": "all", "cullface": "west", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } },
                "east":  { "texture": "all", "cullface": "east", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } }
            }
        }
    ]
}
```
## I know these will get asked, so:
- Can't i write custom model loader?
  - Yes, you can. But you'll have to copy a big part of vanilla model hierarchy and use lots of reflection.
  - If you succeed, you'll not have ability to integrate it with other mods' loaders, and get a lot of `Vanilla model can't have non-vanilla parent`.
  - If you don't trust me, i can show you one, if you want.
- Why capabilities?
  - At first i also thought of not using them, but quickly understood that would write exactly same system. _And who needs duplicate code?_
- Do you already have target use?
  - I do. And i can provide it as example mod if you want me to.

If you have any questions or remarks, drop them here ⬇! Go ahead!

Elix-x committed:
Implemented json model hooks

Implemented json model hooks using capabilities to read data from json
and events to manipulate baked outcome on any level.

williewillus commented:
Usage examples would be nice.

Elix-x commented:
> Example mod provided for very basic, not deep functionality.

[Here](https://github.com/elix-x/MinecraftForge/blob/54d55d98e6f70e4f72d3be7de3027020b071c883/src/test/java/net/minecraftforge/test/JsonModelCapabilityTest.java).
Do you want screenshots of what it does?

Elix-x commented:
Edited example mod details in.

williewillus commented:
I'm not quite sure about using caps for this. 
This seems like 1. having the ability to have a "custom data" section of a vanilla json and 2. Being able to act on this data before baking. 
I feel like in this situation the cap system is just getting in the way.

LexManos commented:
First thing first: Why?
Second thing: Your patch files are horrible, as is your code changes you have a lot of variable renaming and whitespace changing.
Fix that before you do anything else, the PR should be ONLY your INTENTIONAL changes.
You also seem to have no regard for binary compatibility.

And back to the first: I'm not a fan of overly changing the vanilla model format. We did a SIMPLE change to it to allow for it to calculate the permutations.
Why do you need to do extra fancy stuff in code DYNAMICALLY?
If you NEED a functionality that doesnt exist in it, we can discuss those functionalities and see what would be the best.
This is the best way to go about it if you want to actually see the json format improve. And be standardized.
I'm REALLY against building a system that does nothing but encourages people to create non-standard json formats...

Elix-x commented:
@williewillus 2 things:
1. At first, i did try to avoid using capabilities and write other system instead. So we need to deserealize data (_json_), attach it to object (_model_), pass it from one object copy to another by copying (so changes to one don't affect another) and we need to be able to retrive this data whenever we do stuff to affect it (_baking_). Seems familiar?
2. We can affect outcome prior to baking, during baking and post baking.

@LexManos 

> Second thing: Your patch files are horrible, as is your code changes you have a lot of variable renaming and whitespace changing.
> Fix that before you do anything else, the PR should be ONLY your INTENTIONAL changes.
> You also seem to have no regard for binary compatibility.

First, i checked 3 times for bad whitespaces (and checked again). I don't see any that are bad (there are new lines between caps methods, but they are there in all other caps patches).
Second, i did not change any variables' names, i simply ran `genPatches`. It seems because i added new arguments to methods, patch generator renamed them all. This can be fixed manually though (i will).

Elix-x committed:
Fixed binary compatibility & patches

- Fixed binary compatibility.
- Decreased size of patches.
- Fixed bad whitespaces.

Elix-x committed:
Added better examples

Better examples: one for model level and one for vertex level.

Elix-x commented:
@LexManos fixed everything!

& @williewillus added more complex examples (and ones i actually need):

---

## Examples

I included 2 examples with this PR. One for model level modification and another one for vertex level modification.
[Here's example mod with capabilities and events](https://github.com/elix-x/MinecraftForge/blob/8eb94cf1af76a713610a3cc9e86ace6a8fa8ec24/src/test/java/net/minecraftforge/test/JsonModelCapabilityTest.java), [here the model file](https://github.com/elix-x/MinecraftForge/blob/8eb94cf1af76a713610a3cc9e86ace6a8fa8ec24/src/test/resources/assets/jsonmodelcapabilitytest/models/block/testblock.json) and here the result:
![Fancy](http://i.imgur.com/Llojd7z.png)

_**Note: these are the actual functionalities i need for a mod.**_

### Include:

Includes all quads of another model into this one with possible transformations. Does not require a ton of copy-pasta for repeating complex models.
Result:
![Bacon!](http://i.imgur.com/j80IHDf.png)
Code (responsible parts of code of example mod, cut out for readability):

``` java
    @CapabilityInject(Include.class)
    private static final Capability<Include> include = null;
...
    CapabilityManager.INSTANCE.register(Include.class, new Include.Storage(), Include.class);
...
    @SubscribeEvent
    public void attachCapsModel(ClientAttachCapabilitiesEvent.ModelBlock event){
        event.addCapability(new ResourceLocation("test:include"), new ICapabilitySerializable<NBTBase>(){...});
    }
...
    @SubscribeEvent
    public void include(VanillaModelWrapperEvent.Bake.Model.Post event){
        if(event.getModel().hasCapability(include, null)){
            ...
        }
    }
...
    public static class Include {

        List<Pair<String, ComplexTransforms>> include = new ArrayList<Pair<String,ComplexTransforms>>();

        static class Storage implements IStorage<Include> { ... }

        }
    public static class ComplexTransforms
    {

        List<Transform> transforms = new ArrayList<Transform>();

        ...

        static class Transform {

            private Vector3f translation = null;
            private Vector3f rotation = null;
            private Vector3f scale = null;

            ...

        }
    }
```

Example json:

``` json
{
    "capabilities":{
        "test:include": [
            {
                "model": "block/furnace",
                "transform": [
                    {
                        "rotation": {
                            "x": 0,
                            "y": 0,
                            "z": 180
                        },
                        "scale": {
                            "x": 0.5,
                            "y": 0.5,
                            "z": 0.5
                        }
                    }
                ]
            },
            {
                "model": "block/beacon",
                "transform": [
                    {
                        "scale": {
                            "x": 0.5,
                            "y": 0.5,
                            "z": 0.5
                        }
                    },
                    {
                        "translation": {
                            "x": 0,
                            "y": 1,
                            "z": 0
                        }
                    }
                ]
            },
            {
                "model": "block/end_rod",
                "transform": [
                    {
                        "translation": {
                        "x": 0,
                            "y": 1.25,
                            "z": 0
                        }
                    }
                ]
            }
        ]
    }
}
```

### UVLightmap

We have `shade`, but we can't cofigure intensity.
Result:
![LEDs!](http://i.imgur.com/bKEHaF9.png)
Code (responsible parts of code of example mod, cut out for readability):

``` java
    @CapabilityInject(UVLightmap.class)
    private static final Capability<UVLightmap> uvlightmap = null;
...
    CapabilityManager.INSTANCE.register(UVLightmap.class, new UVLightmap.Storage(), UVLightmap.class);
...
    @SubscribeEvent
    public void attachCapsBlockPartFace(ClientAttachCapabilitiesEvent.BlockPartFace event){
        event.addCapability(new ResourceLocation("test:uvlightmap"), new ICapabilitySerializable<NBTBase>() { ... });
    }
...
    @SubscribeEvent
    public void uvlightmap(BakeBlockPartFaceEvent event){
        if(event.getFace().hasCapability(uvlightmap, null)){
            ...
        }
    }
...
    public static class UVLightmap {

        float block = -1;
        float sky = -1;

        static class Storage implements IStorage<UVLightmap> {

            ...

        }

    }
```

Example json:

``` json
{
    "textures":{
        "all": "jsonmodelcapabilitytest:blocks/lights"
    },
    "elements": [
        {   
            "from": [ 2, 2, 2 ],
            "to": [ 14, 14, 14 ],
            "faces": {
                "down":  { "texture": "all", "cullface": "down", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.007, "sky": 0.007 } } },
                "up":    { "texture": "all", "cullface": "up", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.007, "sky": 0.007 } } },
                "north": { "texture": "all", "cullface": "north", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.004, "sky": 0.004 } } },
                "south": { "texture": "all", "cullface": "south", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } },
                "west":  { "texture": "all", "cullface": "west", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } },
                "east":  { "texture": "all", "cullface": "east", "uv": [0, 0, 16, 16], "capabilities": { "test:uvlightmap": { "block": 0.001, "sky": 0.001 } } }
            }
        }
    ]
}
```

williewillus commented:
I like what this can allow to happen but I still think using caps for it in code is unnecessary. Just make a simple parsing system, see how custom LootFunctions and LootConditions are defined for an example.

Elix-x commented:
@williewillus Well, before model is actually baked, it's being modified and cloned a lot of times. So we need to add structure into model that can be deep copied from one copy to another so changes to one clone don't affect another.

kashike commented on src/main/java/net/minecraftforge/client/event/BakeBlockPartFaceEvent.java:
```diff
@@ -0,0 +1,101 @@
+package net.minecraftforge.client.event;
```
Missing license header.

Elix-x replied:
Ok, will add to all new classes.

Elix-x committed:
Added license headers

Added license headers to new classes.

Elix-x commented:
@kashike added to all new classes inside forge.

Any more questions / concerns?

Actuarius commented:
@williewillus added labels [Feature]

Elix-x commented:
Any questions?

Elix-x commented:
@RainWarrior any updates?

Elix-x commented:
Almost 2 moths have passed. Any updates?

mezz commented:
I think this change is too massive for anyone to review it thoroughly.
Aside from testing this for correctness, it needs to be tested for performance as well.

There is a great risk of bugs when adding this much code to Forge, and although this seems like an obvious improvement in some ways, it's really hard to justify.

If you can break this up into smaller more incremental PRs, or just add a necessary hook to allow you to use your own library mod instead, I think that would have better potential.

Elix-x commented:
@mezz 
I **can** break it into multiple small PRs each one adding capabilities & events for specific layer of model (model-quads-sides-vertices).
I **can not** put it into my library as it requries modification of vanilla classes (and quite a bit).

Performance impact - 0 during runtime (by itself). These hooks are all for loading phase. They allow modification of model after it's loaded and before it's baked (and uploaded). So runtime is not affected...
Well, of course, unless somebody adds 9000 quads to each model in his mod.

mezz commented:
Is there a way for you to bypass the vanilla classes entirely?
Patching vanilla this heavily is not a sustainable idea, changes from Mojang can nuke it completely.

Elix-x commented:
These might seem heavy vanilla modifications, but there was no PR that introduced capabilities to new things that had less modifications.

Can i bypass vanilla classes entirely? Yes.
Will it work if multiple modders do it? No.
I have answered these in FAQ already. Quoting FAQ:

---
**I know these will get asked, so:**
- Can't i write custom model loader?
  - Yes, you can. But you'll have to copy a big part of vanilla model hierarchy and use lots of reflection.
  - If you succeed, you'll not have ability to integrate it with other mods' loaders, and get a lot of `Vanilla model can't have non-vanilla parent`.
  - If you don't trust me, i can show you one, if you want.
- Why capabilities?
  - At first i also thought of not using them, but quickly understood that would write exactly same system. _And who needs duplicate code?_
- Do you already have target use?
  - I do. And i can provide it as example mod if you want me to.
---

Just FYI - I created this PR only after i implemented models modifications that i needed without ASM modifications. I had to literally shove the entire models structure classes in one file in order to be able to this. Combined with hacking static final fields... You get it.

liach commented:
@Elix-x You should resolve the merge conflict.

Elix-x commented:
@liach That's new.

@mezz So is splitting this PR worth a try?

mezz commented:
I think you should hold off for now. Without fry reviewing, there's not really anyone else to look deeply into this for critique or issues.

Elix-x commented:
@mezz Ok. Found the culprit commit - https://github.com/MinecraftForge/MinecraftForge/commit/6a125ae0d46aa9ee92e80c052b8a1031f829ca8a, should be not hard to fix. Meanwhile, do you know when @RainWarrior will be able to take a look? Because it's kinda been over 4 months.

mezz commented:
Unfortunately I don't know, sorry.

Elix-x committed:
Merge branch '1.11.x' into JsonModelCaps

Elix-x commented:
Resolved conflicts.

Elix-x committed:
Merge branch '1.11.x' into JsonModelCaps

Elix-x committed:
Updated capabilities events to match new standard

Added type params, using new constructors instead.

Elix-x committed:
Returned method required for client caps

Returned gatherCapabilitiesNotNull to event factory (it was deleted by
accident - my workspace bugged and showed it as unused).

Elix-x commented:
Any updates?

williewillus commented:
I still think the capabilities get massively in the way compared to a simple map where you register handlers for a given ResLoc. More than half of the patch is capability boilerplate.

Elix-x commented:
@williewillus It would be so simple, if models weren't rebaked, modified and created over again. With each time a new instance being created. And we do not want to share same objects across different model instances. This can be solved by either using clone or serializing and deserializing the object. Now, we also need to read the object from the JSON in the first place. So using NBT as serialized version seems wise, because this way we can read object from JSON based NBT and clone one.
If we create a system specifically for models, it will pretty much be a copy of capabilities system (and patches would probably be longer).

I'm not saying that caps are ideal here. They are simply the best option of available ones. And @RainWarrior has not said anything yet (even about current implementation)...

Elix-x committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into JsonModelCaps

Conflicts:
	src/main/java/net/minecraftforge/client/ForgeHooksClient.java
	src/main/java/net/minecraftforge/client/model/ItemLayerModel.java
	src/main/java/net/minecraftforge/client/model/ModelLoader.java

Elix-x committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into JsonModelCaps

# Conflicts:
#	src/main/java/net/minecraftforge/client/ForgeHooksClient.java
#	src/main/java/net/minecraftforge/client/model/ItemLayerModel.java
#	src/main/java/net/minecraftforge/client/model/ModelLoader.java

Elix-x committed:
Fixed test mod

Elix-x commented:
Changed target to 1.12 and resolved conflicts.
Any updates?

mezz commented:
Sorry I am closing this. It is a massive change and we don't really have the ability to carefully look it over and test everything thoroughly.
